torch/distributed/fsdp/_optim_utils.py

```diff
@@ -126,17 +131,9 @@ def _communicate_optim_state(
         ConsolidatedOptimState: Consolidated optimizer state for
         ``flat_param``; the state is not populated for non-target ranks.
     """
-    param_index = -1
-    for i, param in enumerate(fsdp_module.params):
-        if param is flat_param:
-            param_index = i
-            break
-    assert param_index >= 0, "`fsdp_module` must own `flat_param`"
-
     state = _ConsolidatedOptimState()
     tensor_state, zero_dim_tensor_state, non_tensor_state = \
         state.tensor_state, state.zero_dim_tensor_state, state.non_tensor_state
-    process_group = fsdp_module.process_group
```

Comment:
We do not have good support for multiple process groups in the module hierarchy. We can readdress if there are such use cases.

`full_optim_state_dict()` now needs to perform a `broadcast_object_list()` (to communicate rank 0's `optim_input` order) and an `all_reduce()` (for validation). These two collectives require _some_ default process group not necessarily tied to one module or parameter. This is why I felt like it would be simpler to just assume a single process group for the entire method.